proof/refine/ARM_HYP/Arch_R.thy

```diff
@@ -2284,18 +2308,14 @@ lemma assoc_invs':
                   split: option.splits if_split_asm)
   by (clarsimp simp: state_hyp_refs_of'_def obj_at'_def projectKOs split: option.splits if_split_asm)
 
-lemma setVCPU_archThreadSet_Some_eq:
-  "do
-    archThreadSet (atcbVCPUPtr_update (\<lambda>_. Just v)) tcb;
-    setObject v $ vcpuTCBPtr_update (\<lambda>_. Just tcb) vcpu;
-    f
-   od = do
-    do
-      archThreadSet (atcbVCPUPtr_update (\<lambda>_. Just v)) tcb;
-      setObject v $ vcpuTCBPtr_update (\<lambda>_. Just tcb) vcpu
-    od;
-    f
-  od" by (simp add: bind_assoc)
+lemma associateVCPUTCB_helper_rewrite:
+  "do archThreadSet (atcbVCPUPtr_update (\<lambda>_. Just v)) tcb;
+      setObject v $ vcpuTCBPtr_update (\<lambda>_. Just tcb) vcpu;
+      f
+   od = do associateVCPUTCB_helper vcpu v tcb;
+           f
+        od"
+  by (simp add: associateVCPUTCB_helper_def bind_assoc)
```

Comment:
Something here I don't get: you're defining associateVCPUTCB_helper to be whatever you want it to be, and then you have this extra rule to fold it up that's just a bit different to your definition. Would it make more sense to define associateVCPUTCB_helper to match exactly what you need it to be and use `fold`?

Reply:
Also, the `Just` is something that comes from Haskell, and should go away with a simp along with the `$`, so it's unusual seeing them here. I'd experiment below in `associateVCPUTCB_invs` with simplifying first before trying a `subst associateVCPUTCB_helper_rewrite[symmetric]`